api/appendix_h.asciidoc

```diff
@@ -371,10 +370,10 @@ When Intermediate Language Programs are not supported:
 | Returns an empty buffer (such as _param_value_size_ret_ equal to `0`) if no devices in the context associated with _program_ support Intermediate Language Programs.
 
 | {clCreateProgramWithIL}
-| Returns {CL_INVALID_VALUE} if no devices in _context_ support Intermediate Language Programs.
+| Returns {CL_INVALID_OPERATION} if no devices in _context_ support Intermediate Language Programs.
 
 | {clSetProgramSpecializationConstant}
-| Returns {CL_INVALID_PROGRAM}, since _program_ cannot have been created from an Intermediate Language.
+| Returns {CL_INVALID_OPERATION} if no devices associated with _program_ support Intermediate Language Programs.
```

Comment:
This PR treats `clGetPipeInfo` and `clSetProgramSpecializationConstant` differently, in the former that the first has two possible return codes and the latter has one.  But they seem similar to me.

Taking possible uses of `clGetPipeInfo` on an implementation that doesn't support pipes:

```c
cl_mem bad_pipe = clCreatePipe(...);
cl_int err1 = clGetPipeInfo(bad_pipe, ...);

// ICD loader case: err1 is definitely CL_INVALID_MEM_OBJECT.
// Non-loader case: err1 is probably CL_INVALID_OPERATION.

cl_mem not_pipe = clCreateBufferWithProperties(...);
cl_int err2 = clGetPipeInfo(not_pipe, ...);

// ICD loader case: Uses 'not_pipe' to reach driver, err2 is probably CL_INVALID_OPERATION.
// Non-loader case: err2 is probably CL_INVALID_OPERATION.
```

So now, equivalently for `clSetProgramSpecializationConstant` on implementations that do not support IL.

```c
cl_program bad_program = clCreateProgramWithIL(...);
cl_int err3 = clSetProgramSpecializationConstant(bad_program, ...);

// ICD loader case: err3 is definitely CL_INVALID_PROGRAM.
// Non-loader case: err3 is probably CL_INVALID_OPERATION.

cl_program not_il = clCreateProgramWithSource(...);
cl_int err4 = clSetProgramSpecializationConstant(not_il, ...);

// ICD loader case: Uses 'not_il' to reach driver, err4 is probably CL_INVALID_OPERATION.
// Non-loader case: err4 is probably CL_INVALID_OPERATION.
```

Both `clGetPipeInfo` and `clSetProgramSpecializationConstant` seem equivalent in the sense that given an OpenCL program that is valid on an implementation that supports the relevant feature, the same program will result in `CL_INVALID_THING` from the ICD loader when backed by an implementation that does not support the relevant feature.

And both allow reaching the implementation anyway via OpenCL programs that will fail on any implementation, due to the multiple ways of getting a `cl_mem` or a `cl_program`.

So should both `clGetPipeInfo` and `clSetProgramSpecializationConstant` should have a similar error code pattern?  Such as:

```suggestion
| Returns either {CL_INVALID_PROGRAM} since _program_ cannot be a valid IL program, or {CL_INVALID_OPERATION} if no devices associated with _program_ support Intermediate Language Programs.
```

Or does the multiple-devices-in-context case make the difference?

Reply:
Ah, this is a very good question.  Thank you for the very thorough write-up!

In short: I think you're right and I think we should remove the `CL_INVALID_MEM_OBJECT` condition from Appendix H.

I had forgotten that a pipe is still a `cl_mem` like any other, so even if you cannot create a `cl_mem` object that is an actual pipe when the implementation doesn't support pipes, you can create a non-pipe `cl_mem` object as shown in your `not_pipe` case, and get to an ICD implementation's `clGetPipeInfo` that way.

This does land us in the situation I was describing offline, where there are two error conditions but we are mandating that one of the two error conditions should be reported:

1. CL_INVALID_MEM_OBJECT if pipe is a not a valid pipe object. (existing)
2. CL_INVALID_OPERATION if pipes are not supported. (new in this PR)

I think that's OK though and it'll be helpful to return `CL_INVALID_OPERATION` consistently.  We will have a CTS test that checks for this behavior.

If this is OK, my action items are:

* Update this PR to remove `CL_INVALID_MEM_OBJECT` from the list of errors in Appendix H.
* I'll also update the error condition for `CL_INVALID_OPERATION` to be a bit more descriptive, so it'll say:
> * `CL_INVALID_OPERATION` if no devices in the context associated with _pipe_ support pipes.
* Update the CTS test to check for `CL_INVALID_MEM_OBJECT` specifically, by using the `not_pipe` case described above.

Reply:
> I think that's OK though and it'll be helpful to return CL_INVALID_OPERATION consistently.

Yes, that sounds okay.

> We will have a CTS test that checks for this behavior.

Based on the main body of specification text saying in a footnote that "The OpenCL specification does not describe the order of precedence for error codes returned by API calls.", technically either error code is still valid, unless appendix H overrides that.  

However, the CTS test sounds fine and in the unlikely event that an implementation fails it then the test can be changed.

> Update the CTS test to check for CL_INVALID_MEM_OBJECT specifically, by using the not_pipe case described above.

Should this say `CL_INVALID_OPERATION`?

Reply:
Great, I've made these changes.

> Should this say `CL_INVALID_OPERATION`?

Yes, it should, thanks!

Reply:
Thanks!